python/examples/token_swap/token_swap.py

```diff
@@ -0,0 +1,370 @@
+    owner_dir_keylet,
+)
+from xrpl_plugin.return_codes import (
+    tecINSUFFICIENT_RESERVE,
+    tecNO_LINE,
+    temBAD_AMOUNT,
+    temBAD_EXPIRATION,
+    tecEXPIRED,
+    temBAD_AMOUNT,
+    tecINSUFFICIENT_FUNDS,
+    tesSUCCESS,
+    tecDIR_FULL,
+)
+
+from xrpl_plugin.sfields import (
+    sf_balance,
+    create_new_sfield,
+    sf_account,
+    sf_expiration,
+    sf_amount,
+    sf_low_limit,
+    sf_owner_count,
+    sf_owner_node,
+)
+from xrpl_plugin.stypes import (
+    STAmount,
+    STAccount,
+    STUInt64,
+    index_hash,
+)
+
+from xrpl_plugin.ledger_objects import (
+    adjust_owner_count,
+    make_sle,
+    ltRIPPLE_STATE,
+)
+
+class TokensNetBalance(InvariantCheck):
+
+    def __init__(self):
+        self.total = 0
+
+    def visit_entry(self, is_delete, before, after):
+        if before is not None and before.get_type() == ltRIPPLE_STATE:
+            self.total += abs(float(after[sf_balance].get_text())) - abs(float(before[sf_balance].get_text()))
+
+    def finalize(self, tx, result, fee, view, j):
+        if tx.get_txn_type() == 54:
+            return self.total == 0
+
+        return True
+
+sf_amount_other = create_new_sfield(STAmount, "AmountOther", 51)
+sf_account_other = create_new_sfield(STAccount, "AccountOther", 52)
+sf_token_swap_id = create_new_sfield(STUInt64, "TokenSwapId", 53)
+
+sfields = [
+    sf_amount_other,
+    sf_account_other,
+    sf_token_swap_id,
+]
+
+ltTOKEN_SWAP_PROPOSE = 0x77
+NEW_TOKEN_SWAP_NAMESPACE = ord("w")
+
+temINVALID_TOKEN_SWAP_ID = -333
+ter_codes = [
+    TERCode(temINVALID_TOKEN_SWAP_ID, "temINVALID_TOKEN_SWAP_ID", "The token swap is incorrect or has already been accepted"),
+]
+
+def new_token_swap_keylet(src, seq):
+    return Keylet(ltTOKEN_SWAP_PROPOSE, index_hash(NEW_TOKEN_SWAP_NAMESPACE, src, seq))
+
+def expired(expiration, now):
+    return expiration <= now.time_since_epoch()
+
+ledger_objects = [
+    LedgerObject(
+        object_type=ltTOKEN_SWAP_PROPOSE,
+        name="TokenSwap",
+        rpc_name="token_swap",
+        object_format=[
+            (sf_account, soeREQUIRED),
+            (sf_amount, soeREQUIRED),
+            (sf_token_swap_id, soeREQUIRED),
+            (sf_amount_other, soeREQUIRED),
+            (sf_account_other, soeREQUIRED),
+            (sf_owner_node, soeREQUIRED),
+            (sf_expiration, soeREQUIRED),
+        ],
+        is_deletion_blocker=True,
+    )
+]
+
+invariant_checks = [TokensNetBalance]
+
+amendment = Amendment("featureTokenSwap", True, VoteBehavior.DEFAULT_NO)
+
+amendments = [amendment]
+
+def preflight_propose(ctx):
+
+    print("Token Swap Create test preflight...")
+
+    tx_proposer_send_token = ctx.tx[sf_amount]
+    tx_approver_send_token = ctx.tx[sf_amount_other] 
+    
+    if tx_approver_send_token.is_xrp() or tx_proposer_send_token.is_xrp():
+        return temBAD_AMOUNT
+
+    if not ctx.tx.is_field_present(sf_expiration):
+        return temBAD_EXPIRATION
+
+
+    return preflight2(ctx)
+
```

Comment:
I would consider adding a `preclaim` function that does things like make sure that the account has enough balance and the destination account exists (unless you're okay with currently-unfunded swaps).